src/array.c

```diff
@@ -629,134 +637,309 @@ static size_t limit_overallocation(jl_array_t *a, size_t alen, size_t newlen, si
+    }
+    else {
+        a->data = (char*)a->data + nbdec;
+    }
+    a->offset = newoffs;
+}
+
+STATIC_INLINE void jl_array_del_at_end(jl_array_t *a, size_t idx, size_t dec,
+                                       size_t n)
+{
+    // no error checking
+    // assume inbounds, assume unshared
+    assert(!a->flags.isshared);
+    char *data = (char*)a->data;
+    size_t elsz = a->elsize;
+    size_t last = idx + dec;
+    if (n > last)
+        memmove(data + idx * elsz, data + last * elsz, (n - last) * elsz);
+    if (elsz == 1)
+        data[n - dec] = 0;
     a->nrows -= dec;
+#ifdef STORE_ARRAY_LEN
+    a->length = a->nrows;
+#endif
+}
 
-    // make sure offset doesn't grow forever due to deleting at beginning
-    // and growing at end
-    size_t newoffs = offset;
-    if (offset >= 13*a->maxsize/20) {
-        newoffs = 17*(a->maxsize - a->nrows)/100;
+JL_DLLEXPORT void jl_array_del_at(jl_array_t *a, ssize_t idx, size_t dec)
+{
+    size_t n = jl_array_nrows(a);
+    size_t last = idx + dec;
+    if (__unlikely(idx < 0))
+        jl_bounds_error_int((jl_value_t*)a, idx + 1);
+    if (__unlikely(last > n))
+        jl_bounds_error_int((jl_value_t*)a, last);
+    // The unsharing needs to happen before we modify the buffer
+    if (__unlikely(a->flags.isshared))
+        array_try_unshare(a);
+    if (idx < n - last) {
+        jl_array_del_at_beg(a, idx, dec, n);
     }
-#ifdef _P64
-    while (newoffs > (size_t)((uint32_t)-1)) {
-        newoffs = newoffs/2;
+    else {
+        jl_array_del_at_end(a, idx, dec, n);
     }
+}
+
+JL_DLLEXPORT void jl_array_del_beg(jl_array_t *a, size_t dec)
+{
+    size_t n = jl_array_nrows(a);
+    if (__unlikely(dec > n))
+        jl_bounds_error_int((jl_value_t*)a, dec);
+    if (__unlikely(a->flags.isshared))
+        array_try_unshare(a);
+    jl_array_del_at_beg(a, 0, dec, n);
+}
+
+JL_DLLEXPORT void jl_array_del_end(jl_array_t *a, size_t dec)
+{
+    size_t n = jl_array_nrows(a);
+    if (__unlikely(n < dec))
+        jl_bounds_error_int((jl_value_t*)a, 0);
+    if (__unlikely(a->flags.isshared))
+        array_try_unshare(a);
+    jl_array_del_at_end(a, n - dec, dec, n);
+}
+
+JL_DLLEXPORT void jl_array_sizehint(jl_array_t *a, size_t sz)
+{
+    size_t n = jl_array_nrows(a);
+    if (sz <= n)
+        return;
+    size_t inc = sz - n;
+    jl_array_grow_end(a, inc);
+    a->nrows = n;
+#ifdef STORE_ARRAY_LEN
+    a->length = n;
 #endif
-    if (newoffs != offset) {
-        size_t anb = a->nrows*es;
-        size_t delta = (offset - newoffs)*es;
-        a->data = (char*)a->data - delta;
-        memmove(a->data, (char*)a->data + delta, anb);
+}
+
+JL_DLLEXPORT jl_array_t *jl_array_copy(jl_array_t *ary)
+{
+    jl_array_t *new_ary = _new_array(jl_typeof(ary), jl_array_ndims(ary),
+                                     &ary->nrows);
+    memcpy(new_ary->data, ary->data, jl_array_len(ary) * ary->elsize);
+    return new_ary;
+}
+
+JL_DLLEXPORT jl_array_t *jl_array_vcat_vectors(jl_value_t *arrays)
```

Comment:
Not really for `vcat` and `copy`. The problem is not about the `memcpy`/`memmove` functions itself but what we do before it. It is not safe to use `unsafe_copy!` on a pointer array as long as we have any allocation after the array itself is allocated.

I agree that having a gc-unsafe region concept in julia is useful. However, for it to be remotely useful, we need to be able to statically analyze what can allocate and what can't (also what can iteract with GC and what can't). The `vcat` and `copy` case can also be more tricky since we will need a way to express that no allocation or safe points are allowed when we return from `jl_new_array` but we don't want to disable the GC before calling `jl_new_array` (or `Array{T}(n)`) (the C code is written to make sure the array is young when `jl_new_array` returns).